frontend/src/components/PhoneNumberInput/PhoneNumberInput.tsx

```diff
@@ -0,0 +1,304 @@
+      ...props
+    },
+    ref,
+  ) => {
+    const styles = useMultiStyleConfig('PhoneNumberInput', props)
+
+    // Internal states of the component.
+    const [inputValue, setInputValue] = useState(propsValue ?? '')
+    const [country, setCountry] = useState(defaultCountry)
+
+    // Refs of the phone number input so focus can be passed to the input when
+    // the selected country changes.
+    const innerInputRef = useRef<HTMLInputElement | null>(null)
+    // Used so any forwarded refs passed can be merged with internal ref.
+    const inputRef = useMergeRefs(innerInputRef, ref)
+
+    const formatter = useMemo(() => new AsYouType(country), [country])
+
+    const inputPlaceholder = useMemo(() => {
+      if (examplePlaceholder === 'off') {
+        return props.placeholder
+      }
+
+      const exampleNumber = getExampleNumber(
+        country,
+        examples,
+      )?.formatNational()
+
+      if (examplePlaceholder === 'aggressive') {
+        return exampleNumber ?? props.placeholder
+      }
+
+      return props.placeholder ?? exampleNumber
+    }, [country, examplePlaceholder, examples, props.placeholder])
+
+    const onInputChange = useCallback(
+      (newValue: string) => {
+        if (inputValue === newValue) return
+
+        // The as-you-type formatter only works with append-only inputs.
+        // Changes other than append require a reset.
+        const isAppend =
+          newValue.length > inputValue.length &&
+          newValue.slice(0, inputValue.length) === inputValue
+
+        if (isAppend) {
+          const appended = newValue.slice(inputValue.length)
+          setInputValue(formatter.input(appended))
+
+          const number = formatter.getNumber()
+          if (number?.country && number.country !== country) {
+            setCountry(number.country)
+          }
+        } else {
+          // Reset the formatter, but do not reformat.
+          // Doing so now will cause the user to loose their cursor position
+          // Wait until blur or append to reformat.
+          formatter.reset()
+          formatter.input(newValue)
+          setInputValue(newValue)
+        }
+
+        const number = formatter.getNumber()
+
+        const e164 = number?.number as string | undefined
+        onChange(e164)
+
+        // On a similar vein, do not set country even if the country has changed
+        // so that the cursor position does not get lost.
+        // Change country on blur instead.
+      },
+      [country, formatter, inputValue, onChange],
+    )
+
+    const onCountryChange = useCallback(
+      (newCountry: CountryCode) => {
+        if (country === newCountry) return
+        onInputChange('')
+        setCountry(newCountry)
+        innerInputRef?.current?.focus()
+      },
+      [country, onInputChange],
+    )
+
+    const onInputBlur = useCallback(() => {
+      const number = formatter.getNumber()
+      onBlur?.()
+
+      onChange(number?.number as string | undefined)
```

Comment:
could i clarify on why this `onChange` is needed, as we would have called it up in the `onInputChange` block already?

Reply:
The number may be formatted again.

This can happen in the below scenario:

1. `onChange` gets called when user types for example "65aabvcd123"
2. `formatter.getNumber().number` will transform that into "65" and cut out the remaining characters since the remaining string is not a valid number
3. Will need to call onChange on this new number.

I will add a new comment for clarity